custom_decorator/activity_utils.py

```diff
@@ -13,28 +13,26 @@ def auto_heartbeater(fn: F) -> F:
     # available via our wrapper, so we use the functools wraps decorator
     @wraps(fn)
     async def wrapper(*args, **kwargs):
-        done = asyncio.Event()
+        tasks = [asyncio.create_task(fn(*args, **kwargs))]
         # Heartbeat twice as often as the timeout
         heartbeat_timeout = activity.info().heartbeat_timeout
         if heartbeat_timeout:
-            asyncio.create_task(
-                heartbeat_every(heartbeat_timeout.total_seconds() / 2, done)
+            tasks.append(
+                asyncio.create_task(
+                    heartbeat_every(heartbeat_timeout.total_seconds() / 2)
+                )
             )
-        try:
-            return await fn(*args, **kwargs)
-        finally:
-            done.set()
+
+        await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
```

Comment:
Won't the heartbeater run forever?

From the PR description you have:

> If the wrapped user Task completes first, the heartbeat_every() asyncio.Task receives a cancel and ends.

But at https://docs.python.org/3/library/asyncio-task.html#asyncio.wait, it says:

> Unlike [wait_for()](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for), wait() does not cancel the futures when a timeout occurs.

So that means just because one of the tasks here completes, the other doesn't. You can take the pending results and cancel them here though I think and then await them.

Reply:
Yes! I saw that and that documentation is SUPER confusing. However, it pertains *only* to the `timeout=` kwarg.  When the return_when= condition is satisfied however, the remaining tasks are Task.cancel()-ed (provided that return_when=ALL_COMPLETED or FIRST_EXCEPTION)

```python
coroutine asyncio.wait(aws, *, timeout=None, return_when=ALL_COMPLETED)
```

Here's a convoluted example:

```python
import asyncio


async def fine():
    try:
        await asyncio.sleep(2)
        print('Everything is happy')
    except asyncio.CancelledError as e:
        print('fine() was cancelled.  Everything is /not/ fine.')
        raise e


async def fail(task_to_cancel=None):
    try:
        await asyncio.sleep(1)
    except asyncio.CancelledError:
        print('fail() was cancelled before it could do its thing')
        raise

    if task_to_cancel:
        print('fail() will cancel fine()')
        task_to_cancel.cancel()
    else:
        print('fail() will raise exc')
        raise Exception('Kaboom!')


async def run():
    print('Start')
    fine_task=asyncio.create_task(fine())
    await asyncio.wait(
        [
            fine_task,
            # asyncio.create_task(fail())
            asyncio.create_task(fail(task_to_cancel=fine_task))
        ],
        return_when=asyncio.FIRST_COMPLETED
    )
    print('run() was able to finish without a try/except')
    print('End')

async def main():
    await run()

if __name__ == '__main__':
    asyncio.run(main())
```

The output of `print('run() was able to finish without a try/except')` shows that the caller of asyncio.wait( ..., return_when=asyncio.FIRST_COMPLETED) continues execution in the event of raised cancelations.  It also shows that the caller need not explicitly cleanup / cancel the remaining tasks....asyncio.wait() does this automatically.

```
Start
fail() will cancel fine()
fine() was cancelled.  Everything is /not/ fine.
run() was able to finish without a try/except
End
```

Reply:
:+1: If you have confirmed _both_ tasks receive the cancel, that works for me. Feel free to mark this PR ready for review when you feel good about it.

Reply:
@cretz Sorry, I may have spoken incorrectly.  There's enough difference between my own tests versus the nice setup in `custom_decorator/` that with the latter, indeed heartbeat_every() does run forever even after `wait_for_cancel_activity()` is canceled.

In short, I'll have more changes & testing to do